tree/tree/src/TTree.cxx

```diff
@@ -5452,7 +5452,7 @@ Int_t TTree::GetEntry(Long64_t entry, Int_t getall)
 
 #ifdef R__USE_IMT
    // At most one parallel read with a single branch
-   unsigned int nSortedBranches(2);
+   unsigned int nSortedBranches = GetListOfBranches()->GetEntries();
```

Comment:
Actually the sorted branches can be less than the total number of branches. The ones that are sorted are to be processed in parallel, but there can be leaf count branches that have to be processed sequentially.
So even if we check here for the total number of branches and we get e.g. 2, it might be one leaf count branch plus one normal branch, so we will end up doing one sequentially and spawning a task for the other one.

Reply:
That said, I believe the logic involving `nSortedBranches` here is not really useful, the variable itself could be removed as it does not have any actual effect.

Reply:
@etejedor thanks for taking a look.
the aim of this PR is precisely to make that useless logic useful.
according to @dpiparo the intention here was to avoid the parallel path if the number of branches is smaller than 2, hence my fix.
can you suggest what is the right value to check instead?

Consider the next lines look like
```c++
if (nSortedBranches > 1 && ROOT::IsImplicitMTEnabled() && fIMTEnabled && !TTreeCacheUnzip::IsParallelUnzip()) {
       if (fSortedBranches.empty()) {
         InitializeBranchLists(true);
         nSortedBranches = fSortedBranches.size();
       }
...
```
so it seems `fSortedBranches.GetEntries()` would not be a good initial value for `nSortedBranches`

Reply:
The intention of Danilo's lines was to fix a bug when triggering parallelism: before, `fSortedBranches.size()` was checked before it was initialized, so its size was 0 and the code didn't go parallel.
We can check the total number of branches here as an indicator for going parallel or not. What I wanted to say with my previous point is that, even if the number of total branches is greater than 1, it is not a guarantee that you will be able to parallelise them. To be sure, we would need to initialize `fSortedBranches` before the `if` and check the size of that list, but we do not want to add that extra overhead to the sequential case.

Reply:
Bottom line: we have two options
- We remove completely the `nSortedBranches` variable
- We add another variable `nTotalBranches` and check on that, even if it is not the perfect indicator.
Anyway:
- Is it that bad that we spawn a task when there is only one branch?
- If you only have one branch, you should not activate this IMT case in the first place

Reply:
@etejedor I see, then is my change ok, would you prefer a different change, or would you remove the check completely?

Reply:
@etejedor I see, thanks for the explanation.

@etejedor @dpiparo @pcanal let me know what you prefer of the two options listed above :smile: